uproot4/exceptions.py

```diff
@@ -0,0 +1,80 @@
+import numpy as np
+from uproot4._util import damerau_levenshtein
+
+
+class KeyInFileError(KeyError):
+    """
+    Exception raised by attempts to find ROOT objects in ``TDirectories``
+    or ``TBranches`` in :py:class:`~uproot4.behaviors.TBranch.HasBranches`, which
+    both have a Python ``Mapping`` interface (square bracket syntax to extract
+    items).
+
+    This exception descends from Python's ``KeyError``, so it can be used in
+    the normal way by interfaces that expect a missing item in a ``Mapping``
+    to raise ``KeyError``, but it provides more information, depending on
+    availability:
+
+    * ``because``: an explanatory message
+    * ``cycle``: the ROOT cycle number requested, if any
+    * ``keys``: a list or partial list of keys that *are* in the object, in case
+      of misspelling
+    * ``file_path``: a path (or URL) to the file
+    * ``object_path``: a path to the object within the ROOT file.
+    """
+
+    def __init__(
+        self, key, because="", cycle=None, keys=None, file_path=None, object_path=None
+    ):
+        super(KeyInFileError, self).__init__(key)
+        self.key = key
+        self.because = because
+        self.cycle = cycle
+        self.keys = keys
+        self.file_path = file_path
+        self.object_path = object_path
+
+    def __str__(self):
+        if self.because == "":
+            because = ""
+        else:
+            because = " because " + self.because
+
+        with_keys = ""
+        if self.keys is not None:
+            to_show = None
+            distance = []
+            for key in self.keys:
+                distance.append(damerau_levenshtein(self.key, key))
+            self.keys = np.array(self.keys)[np.argsort(distance)]
```

Comment:
An information-seeking method like `__str__` should not change the state of an object (assign to `self.keys`). I can't imagine any specific errors that this is likely to cause, but it's always something you don't think about. It's good to minimize mutability as much as possible, leaving everything but a few well-advertised attributes immutable, especially for something we like to think of an an inert object, such as an exception instance. You could either solve this by sorting in every `__str__` (performance is not very important here) or by moving the sorting to the constructor.

Also, be aware that this leaves `keys` as a NumPy array, where a Python list would be expected—the contents are strings and are likely to be a small set of them. Things like that end up being "gotchas."

A simple way to solve this would be to do the sort with a [list sorting key](https://docs.python.org/3/howto/sorting.html) instead of a NumPy-like argsort. Actually, since `damerau_levenshtein` takes two arguments, it would make more sense to use a [cmp parameter](https://docs.python.org/3/howto/sorting.html#the-old-way-using-the-cmp-parameter), which is not preferred for some reason, but a `cmp` function can always be turned into a `key` function.